### ISSUE TITLE
Initializing Dict with Literal keys is always a type error

### ISSUE THREAD

**Issue by shamrin** [microsoft/pyright#247]
**Describe the bug**
Initializing `Dict` with `Literal` keys (e.g. `Dict[Literal, int]`) always produces a type error.
 
**To Reproduce**
```python
from typing import Dict
from typing_extensions import Literal

Number = Literal['One', 'Two']

numbers_mapping : Dict[Number, int] = {
    'One': 1,
    'Two': 2
}
```
**Expected behavior**
Expect to type check the above properly, including pointing out that some keys are not specified in the `Literal` (e.g. if `"Three"` key is added to the initial value).

**Screenshots or Code**
<img width="782" alt="image" src="https://user-images.githubusercontent.com/510678/64965688-20ebb600-d8a6-11e9-8bcf-ae0dba7e179f.png">

**VS Code extension or command-line**
Name: Pyright
Id: ms-pyright.pyright
Description: VS Code static type checking for Python
Version: 1.0.62
Publisher: ms-pyright
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=ms-pyright.pyright

**Comment by erictraut** [comment]
Thanks for the suggestion.

I plan to address this, but supporting this will require a somewhat substantial change to the implementation. Currently, pyright always evaluates the type of the righthand side of an assignment and then determines whether it matches the type of the lefthand side. The problem is that when it encounters literal expressions like in your example, it needs to guess what type was intended. In your example, the type of the righthand side could be described as any of the following:

1. Dict[Any, Any]
2. Dict[str, int]
3. Dict[Union[Literal['One', 'Two']], int]
4. Dict[Union[Literal['One', 'Two']], Union[Literal[1], Literal[2]]]

In this particular case, pyright is guessing #2, but it would need to guess #3 for the RHS to match your LHS type.

The correct fix is to change the way assignment expressions are checked. It would need to first evaluate the expected type of the LHS and then use that as a hint to determine whether the literal value in the RHS matches it. This is the technique TypeScript uses to handle cases like this.

**Comment by erictraut** [comment]
I've added the support to address this issue, and it will be included in the next published version of pyright. Thanks again for the bug report.

**Comment by erictraut** [comment]
This is now fixed in pyright 1.0.63, which I just published.
